homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,226 @@
+        return self.heatpump.get_coils()
+
+    @property
+    def unique_id(self) -> str:
+        """Return unique id for this coordinator."""
+        return self.config_entry.unique_id or self.config_entry.entry_id
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information for the main device."""
+        return DeviceInfo(identifiers={(DOMAIN, self.unique_id)})
+
+    def get_coil_value(self, coil: Coil) -> int | str | float | None:
+        """Return a coil with data and check for validity."""
+        if coil := self.data.get(coil.address):
+            return coil.value
+        return None
+
+    def get_coil_float(self, coil: Coil) -> float | None:
+        """Return a coil with float and check for validity."""
+        if value := self.get_coil_value(coil):
+            return float(value)
+        return None
+
+    async def async_write_coil(
+        self, coil: Coil | None, value: int | float | str
+    ) -> None:
+        """Write coil and update state."""
+        if not coil:
+            raise HomeAssistantError("No coil available")
+
+        coil.value = value
+        coil = await self.connection.write_coil(coil)
+
+        if self.data:
+            self.data[coil.address] = coil
+            self.async_update_listeners()
+
+    async def _async_update_data(self) -> dict[int, Coil]:
+        @retry(
+            retry=retry_if_exception_type(CoilReadException), stop=stop_after_attempt(2)
+        )
+        async def read_coil(coil: Coil):
+            return await self.connection.read_coil(coil)
+
+        callbacks: dict[int, list[CALLBACK_TYPE]] = {}
+        for update_callback, context in list(self._listeners.values()):
+            assert isinstance(context, set)
+            for address in context:
+                callbacks.setdefault(address, []).append(update_callback)
+
+        result: dict[int, Coil] = {}
+
+        for address, callback_list in callbacks.items():
+            try:
+                coil = self.heatpump.get_coil_by_address(address)
+                self.data[coil.address] = result[coil.address] = await read_coil(coil)
+            except (CoilReadException, RetryError) as exception:
+                self.logger.warning("Failed to update: %s", exception)
+            except CoilNotFoundException as exception:
+                self.logger.debug("Skipping missing coil: %s", exception)
+
+            for update_callback in callback_list:
+                update_callback()
+
+        return result
+
+
+class CoilEntity(CoordinatorEntity[Coordinator]):
+    """Base for coil based entities."""
+
+    _attr_has_entity_name = True
+    _attr_entity_registry_enabled_default = False
+
+    def __init__(
+        self, coordinator: Coordinator, coil: Coil, entity_format: str
+    ) -> None:
+        """Initialize base entity."""
+        super().__init__(coordinator, {coil.address})
+        self.entity_id = async_generate_entity_id(
```

Comment:
Why do we need a different entity id than entity name by default?

Reply:
The "name" is a friendly name. We have a better machine name which is close to friendly but more complete.

Reply:
But is it really needed to use the machine name? The users can update their names and entity ids anyway.

Reply:
It's easier to know which coil it is, since this will include the parameter number, which is what is normally used to refer to nibe parameters in all documentation and help.

Reply:
Ok.